### ISSUE TITLE
Error: Cannot find module 'abbrev'

### ISSUE THREAD

**Issue by miktam** [tjunnone/npm-check-updates#256]
When trying to upgrade to the newest version:

```
➜  npm install -g npm-check-updates
/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib
└── (empty)

npm ERR! Darwin 15.5.0
npm ERR! argv "/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/bin/node" "/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/bin/npm" "install" "-g" "npm-check-updates"
npm ERR! node v4.4.4
npm ERR! npm  v3.8.9
npm ERR! path /Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib/node_modules/.staging/abbrev-b3f03229
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib/node_modules/.staging/abbrev-b3f03229' -> '/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib/node_modules/npm-check-updates/node_modules/npm/node_modules/abbrev'
npm ERR! enoent ENOENT: no such file or directory, rename '/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib/node_modules/.staging/abbrev-b3f03229' -> '/Users/andreikarpushonak/.nvm/versions/node/v4.4.4/lib/node_modules/npm-check-updates/node_modules/npm/node_modules/abbrev'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/andreikarpushonak/Repos/payment-service/npm-debug.log
npm ERR! code 1

➜  ncu --version
module.js:327
    throw err;
    ^

Error: Cannot find module 'abbrev'
    at Function.Module._resolveFilename (module.js:325:15)
```

**Comment by vadirn** [comment]
Reinstall did help.

**Comment by cristian-sima** [comment]
Go the same. I reinstall it. Has problems. I reinstalled Node.js and it works for `npm-check-updates`.

But I have a problem with `ncu`. When I type that I get:

```
module.js:442
    throw err;
    ^

Error: Cannot find module 'C:\Program Files\nodejs\node_modules\npm-check-updates\bin\ncu'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.runMain (module.js:575:10)
    at run (bootstrap_node.js:352:7)
    at startup (bootstrap_node.js:144:9)
    at bootstrap_node.js:467:3
```

**Comment by raineorshine** [comment]
@miktam npm-check-updates does not use the `abbrev` module, so this must be either an issue with a dependency or another example of the many ways that npm fails to install modules correctly.

@christian-sima That looks like an install or permissions issue. It is not an issue with npm-check-updates itself. If you are still experiencing the problem after re-installing and troubleshooting, please open a separate issue and we'll see what we can do, although it is likely an issue specific to your npm installation.

**Comment by miktam** [comment]
thank you, reinstall did help:

```
npm uninstall -g npm-check-updates
npm install -g npm-check-updates
```

**Comment by Beingbook** [comment]
It's strange. I still can't use this package because `abbrev` module

**Comment by louh** [comment]
Did you try uninstall and re-installing, @Beingbook ? I just discovered this issue when facing the same problem, and that helped me.

**Comment by cristian-sima** [comment]
It works for `npm-check-updates` but it does not work for `ncu`.


 How do I set back the alias for `ncu` ?

**Comment by Beingbook** [comment]
I uninstalled and installed `npm-check-updates` again but it throws:
```sh
Maniaxui-MacBook-Pro:telpo-app maniax$ npm-check-updates
[INFO]: You can also use ncu as an alias
module.js:341
    throw err;
    ^

Error: Cannot find module 'abbrev'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/maniax/.npm-packages/lib/node_modules/npm-check-updates/node_modules/npm/node_modules/nopt/lib/nopt.js:10:14)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/maniax/.npm-packages/lib/node_modules/npm-check-updates/node_modules/npm/lib/config/defaults.js:8:12)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
Maniaxui-MacBook-Pro:telpo-app maniax$ ncu
module.js:341
    throw err;
    ^

Error: Cannot find module 'abbrev'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/maniax/.npm-packages/lib/node_modules/npm-check-updates/node_modules/npm/node_modules/nopt/lib/nopt.js:10:14)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/maniax/.npm-packages/lib/node_modules/npm-check-updates/node_modules/npm/lib/config/defaults.js:8:12)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
Maniaxui-MacBook-Pro:telpo-app maniax$ 
```

**Comment by Beingbook** [comment]
It just fixed after using node 6.3 (before was 5.11)

**Comment by cristian-sima** [comment]
I am using `6.3.0`

**Comment by cristian-sima** [comment]
I am installing `v6.3.1` to see if that solves the problem

**Comment by cristian-sima** [comment]
Still the same problem

![image](https://cloud.githubusercontent.com/assets/12199042/17170085/42a7b0e0-53f3-11e6-8d1f-d6256bd80e89.png)

**Comment by raineorshine** [comment]
This looks like a system PATH/permissions issue.

**Comment by cristian-sima** [comment]
I have `windows` (I know, I hate it). I have tried also as Administrator => no result


There is no permission issue

**Comment by raineorshine** [comment]
If you go into the file system, is `C:\Program Files\nodejs\node_mdules\npm-check-updates\bin\ncu` there? If not, is it just `ncu` that is missing or are other parts of that path missing?

**Comment by cristian-sima** [comment]
I did not have anything in that folder.

It works now, by installing locally in that director the ncu

**Comment by raineorshine** [comment]
I'm not sure why it failed to copy in the install. `/bin/ncu` should look like [this file](https://github.com/tjunnone/npm-check-updates/blob/master/bin/ncu).

Sorry for the trouble! If anyone has any idea why this might be occurring for some people and not others, please report here. If we reproduce the issue across systems we can troubleshoot more easily.

**Comment by kevinSuttle** [comment]
This worked for me: 
```
nvm install stable (6.3.1)
npm uninstall -g npm-check-updates
npm install -g npm-check-updates
```

**Comment by steve-gray** [comment]
I can confirm the same issue. Was running 6.1, but updated to 6.3.1 - no joy: the error happened every time I tried to update. Running on OSX El-Cap with latest updates.

However... When I did npm uninstall and then npm install as per @kevinSuttle's suggestion - mysteriously it worked. NFI why a straight reinstall over the top chokes....

**Comment by johnnyreilly** [comment]
I'm running on Windows with node v4.5.0 and npm v3.10.8.  I've tried all the suggestions in this thread but I'm afraid I'm still experiencing the following error:

```
C:\source\proverb-signalr-react [master +2 ~3 -11 !]> npm-check-updates
[INFO]: You can also use ncu as an alias
module.js:327
    throw err;
    ^

Error: Cannot find module 'abbrev'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (C:\Program Files\nodejs\node_modules\npm-check-updates\node_modules\npm\node_modules\nopt\lib
\nopt.js:10:14)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (C:\Program Files\nodejs\node_modules\npm-check-updates\node_modules\npm\lib\config\defaults.j
s:8:12)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
```

**Comment by raineorshine** [comment]
@johnnyreilly I'm sorry you're running into issues. I don't have a Windows machine, but maybe someone else on this thread using Windows can assist.

**Comment by memelet** [comment]
This happens on OSX as well with versions 0.23.2 and 0.32.1

$ npm uninstall --global npm-check-updates    # works
$ npm install --global npm-check-updates        # works
$ npm install --global npm-check-updates        # fails
$ npm uninstall --global npm-check-updates    # works
$ npm install --global npm-check-updates        # works
$ npm install --global npm-check-updates        # fails

So even though ncu does not use abbrev, it always fails to install if already installed.

On ubuntu I don't see this problem. (But of course the set global modules is different).

**Comment by memelet** [comment]
Hmm, it seem ncu does indeed use abbrev, no?

```
❯ ncu                                                                                                                 enrich-core/git/master
module.js:440
    throw err;
    ^

Error: Cannot find module 'abbrev'
    at Function.Module._resolveFilename (module.js:438:15)
    at Function.Module._load (module.js:386:25)
    at Module.require (module.js:466:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/bkaplan/.nvm/versions/node/v6.1.0/lib/node_modules/npm-check-updates/node_modules/npm/node_modules/nopt/lib/nopt.js:10:14)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:456:32)
    at tryModuleLoad (module.js:415:12)
    at Function.Module._load (module.js:407:3)
    at Module.require (module.js:466:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/bkaplan/.nvm/versions/node/v6.1.0/lib/node_modules/npm-check-updates/node_modules/npm/lib/config/defaults.js:8:12)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:456%
```

**Comment by memelet** [comment]
Ok, this fails also on Ubuntu in the same way as with OSX

**Comment by raineorshine** [comment]
@memelet Thanks, it is helpful to know that it fails to install only after it is already installed. I wonder what the issue could be? It appears to be an issue with either the `nopt` package or `npm` itself.

`ncu` does not use `abbrev` directly, as can be seen in the [package.json](https://github.com/tjunnone/npm-check-updates/blob/master/package.json). It is used by a dependency though. From your error message, it seems to be npm -> nopt.

**Comment by mathieumg** [comment]
Just gotta be patient now: https://github.com/npm/nopt/pull/60

**Comment by tom-drake** [comment]
If people are still having trouble with this I have some information to share, specifically for Ubuntu 16.04 (it may translate for other flavours of Linux)

I've noticed there is both a node_modules under `/usr/lib` and `/usr/local/lib`. When I was running `ncu` it was failing due to it running under `/usr/local/lib`, but while using the install command it was updating the package in `/usr/lib`. I deleted `npm-check-updates` from `/usr/local/lib` and all is well.

**Comment by gonimar** [comment]
+1

**Comment by wilfil** [comment]
It worked for me:

npm link @angular/cli

**Comment by wilfil** [comment]
It worked for me:

`npm link @angular/cli`

**Comment by frclasso** [comment]
It worked for me too!!

**Comment by liranco4** [comment]
work for me,
Thanks a lot:)

**Comment by raineorshine** [comment]
Closing due to inactivity
